---
title: "beadarray"
author: "Mark Dunning"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
vignette: >
  %\VignetteIndexEntry{Introduction to beadarray}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


# beadarray: R classes and methods for Illumina bead-based data

Mark Dunning, Mike Smith, Johnathan Cairns, Andy Lynch and Matt Ritchie


## Introduction

beadarray is a Bioconductor package for the analysis of Illumina data.  The main feature of the package is its ability to process raw data (text files and TIFFs) from  BeadScan and store this information in convenient R classes.
Access to this bead-level data offers users the choice between different image processing 
and background correction methods and allows for detailed quality assessments to be carried out for 
each array. The summarised intensities bead-summary data output by Illumina's BeadStudio software can 
also be imported using the beadarray package.

The documentation for this package is split into separate user guides: one for bead-level data and 
a second for bead-summary data.
These documents may be accessed using the following commands:

```{r}
library("beadarray")
beadarrayUsersGuide(topic="beadlevel")
beadarrayUsersGuide(topic="beadsummary")
```




## Companion packages to beadarray

In order to reduce the download size of beadarray, we have moved the example dataset to a Bioconductor data package beadarrayExampleData that can be downloaded in the usual way. An expanded dataset and more detailed commentary on the processing of BeadArray data is provided by the BeadArrayUseCases package. Finally, the beadarray package is also able to interact with various Illumina annotation packages in Bioconductor if they are installed. 

```{r eval=FALSE}
install.packages("BiocManager")
BiocManager::install("beadarrayExampleData")

```




## Citing beadarray


- If you use \Rpackage{beadarray} for the analysis or pre-processing of Illumina data please cite

Dunning MJ, Smith ML, Ritchie ME, Tavar\'e S. beadarray: R classes and
methods for Illumina bead-based data}. Bioinformatics, 2007 Aug 15; 23(16):2183-4

- For recommendations on the low-level analysis of bead-level data, please cite

Dunning MJ, Barbosa-Morais NL, Lynch AG, Tavar\'e S, Ritchie ME. Statistical issues in the analysis of Illumina data. 
BMC Bioinformatics. 2008 Feb 6;9(1):85.

For a validation of the VST method (implemented in the lumi package) on the spike-in data set and a comparison 
of the results obtained from a full bead-level analysis, see

Dunning MJ, Ritchie ME, Barbosa-Morais NL, Tavar\'e S, Lynch AG. Spike-in validation of an Illumina-specific 
variance-stabilizing transformation. BMC Research Notes, 2008, 1:18.

If you use BASH to detect spatial artefacts, please cite

If you use the probe reannotation for Illumina expression arrays (used in the current Bioconductor annotation packages named illuminaXXXXX.db) 
please cite

Barbosa-Morais NL, Dunning MJ, Samarajiwa SA, Darot JF, Ritchie ME, Lynch AG, Tavar\'e S. A re-annotation pipeline for Illumina BeadArrays: improving the interpretation of gene expression data.. Nucleic Acids Research 2010 Jan;38(3)e17.Epub 2009 Nov 18.


## Getting help

Wherever possible, please send all queries about beadarray to the
Bioconductor mailing list at bioconductor@stat.math.ethz.ch. This will
help maintain a searchable archive of questions and responses.
When posting to the list, please include the commands you used along with the
version of beadarray and R you are working with.
Version information can be obtained by running the following command:

```{r}
sessionInfo()

```

