%
% NOTE -- ONLY EDIT beadarray.rnw!!!
% beadarray.tex file will get overwritten.
%
%\VignetteIndexEntry{beadarray Vignette}
%\VignetteDepends{}
%\VignetteKeywords{beadarray expression Analysis}
%\VignettePackage{beadarray}
\documentclass{article}

\usepackage{hyperref}
\usepackage{Sweave}
\textwidth=6.2in
\textheight=8.5in
%\parskip=.3cm
\oddsidemargin=.1in
\evensidemargin=.1in
\headheight=-.3in

\newcommand{\Rfunction}[1]{{\texttt{#1}}}
\newcommand{\Robject}[1]{{\texttt{#1}}}
\newcommand{\Rpackage}[1]{{\textsf{#1}}}
\newcommand{\Rmethod}[1]{{\texttt{#1}}}
\newcommand{\Rfunarg}[1]{{\texttt{#1}}}
\newcommand{\Rclass}[1]{{\textit{#1}}}



\newcommand{\classdef}[1]{
  {\em #1}
}
\begin{document}
\title{\Rpackage{beadarray}: R classes and methods for Illumina bead-based data}

\author{Mark Dunning and Matt Ritchie}

\maketitle



\section*{Introduction}

\Rpackage{beadarray} is a Bioconductor package for the analysis of Illumina data. 
The main feature of the package is its ability to process raw data (text files and TIFFs) from BeadScan and store this information in convenient R classes.
Access to this \emph{bead-level data} offers users' the choice between different image processing and background correction methods and allows for detailed quality assessments to be made on each array.
The summarised intensities (\emph{bead-summary data}) output by Illumina's BeadStudio software can also be imported using the \Rpackage{beadarray} package.

The documentation for this package is split into two separate user guides, one for
bead-level data and a second for bead-summary data. 
These documents may be accessed using the following commands:

<<Help, eval=FALSE>>=
library(beadarray)
beadarrayUsersGuide(topic="beadlevel")
beadarrayUsersGuide(topic="beadsummary")
@

\section{Citing \Rpackage{beadarray}}

If you use \Rpackage{beadarray} for the analysis or pre-processing of
Illumina data please cite:\\

Dunning MJ, Smith ML, Ritchie ME, Tavar\'e S. \emph{beadarray: R classes and
  methods for Illumina bead-based data}. Bioinformatics, 2007 Aug 15; 23(16):2183-4. Epub 2007 Jun 22.\\

For recommendations on the low-level analysis of bead-level data, or if you make use of the spike-in data set available from\\

{\tt http://www.compbio.group.cam.ac.uk/Resources/spike/}\\

please cite:\\

Dunning MJ, Barbosa-Morais NL, Lynch AG, Tavar\'e S, Ritchie ME. \emph{Statistical issues in the analysis of Illumina data}. BMC Bioinformatics. 2008 Feb 6;9(1):85.\\

For a validation of the VST method (implemented in the \Rpackage{lumi} package) on the spike-in data set, along with a comparison of the results obtained from a full bead-level analysis, see

Dunning MJ, Ritchie ME, Barbosa-Morais NL, Tavar\'e S, Lynch AG. \emph{Spike-in validation of an Illumina-specific variance-stabilizing transformation}. BMC Research Notes, 2008, 1:18.\\

Probe reannotation for Illumina expression arrays are available from:\\

{\tt http://www.compbio.group.cam.ac.uk/Resources/Annotation/}\\

If you make use of these, please cite:\\

Barbosa-Morais NL, Dunning MJ, Ritchie ME, Lynch AG, Tavar\'e S. \emph{Reannotation of Illumina BeadArray probes improves the interpretation of gene expression data}. (in preparation).

\section{Asking for help on \Rpackage{beadarray}}

Wherever possible, please send all queries about \Rpackage{beadarray} to the
Bioconductor mailing list at {\tt bioconductor@stat.math.ethz.ch}. This will
help maintain a searchable archive of questions and responses.
When posting to the list, please include the commands you used along with the
version of \Rpackage{beadarray} and {\tt R} you are working with.
Version information can be obtained by running the following command:

<<Mailing list, eval=FALSE>>=
sessionInfo()
@

\end{document}
